### PR TITLE
feat: add getMsbD_sshiftRight, getMsbD_sshiftRight', getLsbD_sshiftRight, getMsbD_ushiftRight

### PR DESCRIPTION
**Pull request by luisacicolini**

### CODE REVIEW

File: SSA/Projects/InstCombine/ForLean.lean [hunk 1]
```diff
@@ -543,6 +543,122 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+    simp
+  case succ n ih =>
+  -- if i remove the only here everything breaks. why?
+    simp only [gt_iff_lt, lt_add_iff_pos_left, add_pos_iff, zero_lt_one, or_true,
+      ↓reduceIte, ih, BitVec.msb, getMsbD_ushiftRight, Bool.and_false]
+
+-- this one exists already under the name : sshiftRight_msb_eq_msb
+-- theorem msb_sshiftRight {x : BitVec w} {n : Nat} :
+--     (x.sshiftRight n).msb = x.msb := by
+--   rw [sshiftRight_msb_eq_msb]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Should we then rename the other one to msb_sshiftRight and introduce a deprecated annotation.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 2]
```diff
@@ -543,6 +543,122 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+
+theorem msb_ushiftRight {x : BitVec w} {n : Nat} :
+    (x.ushiftRight n).msb = if n > 0 then false else x.msb := by
+  induction n
+  case zero =>
+    simp
+  case succ n ih =>
+  -- if i remove the only here everything breaks. why?
+    simp only [gt_iff_lt, lt_add_iff_pos_left, add_pos_iff, zero_lt_one, or_true,
+      ↓reduceIte, ih, BitVec.msb, getMsbD_ushiftRight, Bool.and_false]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: You probably write into the wrong direction with some lemma. Would be useful to identify which one that is.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 3]
```diff
@@ -543,6 +543,122 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+
+theorem getMsbD_ushiftRight {x : BitVec w} {i n : Nat} :
+    getMsbD (x.ushiftRight n) i = (decide (i < w) && if i < n then false else getMsbD x (i - n)) := by
+  simp only [getMsbD, Bool.if_false_left]
+  by_cases h : i < w
+  · simp [h]
+    by_cases h₁ : i < n
+    · simp [h₁]
+      rw [BitVec.getLsbD_ge]
+      omega
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
      rw [BitVec.getLsbD_ge (by omega)]
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 4]
```diff
@@ -543,6 +543,122 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+      · have h₃ : (i < y.toNat) := by omega
+        simp [h₂, h₃]
+  · simp [h]
+
+theorem getMsbD_ushiftRight {x : BitVec w} {i n : Nat} :
+    getMsbD (x.ushiftRight n) i = (decide (i < w) && if i < n then false else getMsbD x (i - n)) := by
+  simp only [getMsbD, Bool.if_false_left]
+  by_cases h : i < w
+  · simp [h]
+    by_cases h₁ : i < n
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
  · simp [h]
    by_cases h₁ : i < n
```

Non-terminal simp.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 5]
```diff
@@ -543,6 +543,122 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+    · simp [h₁]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Non-terminal simp.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 6]
```diff
@@ -543,6 +543,122 @@ theorem shiftLeft_and_distrib' {x y : BitVec w} {n m : Nat} :
...
+      rw [BitVec.getLsbD_ge]
+      omega
+    · have h₂ : (i - n < w) := by omega
+      simp [h₁, h₂]
+      congr
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Non-terminal simp?

File: SSA/Projects/InstCombine/ForLean.lean (hunk 1 above)
Comment (luisacicolini) [reply SSA/Projects/InstCombine/ForLean.lean]: right, thanks! shall we do this directly when upstreaming this to lean?

File: SSA/Projects/InstCombine/ForLean.lean (hunk 3 above)
Comment (luisacicolini) [reply SSA/Projects/InstCombine/ForLean.lean]: `by omega` does not seem to work here unfortunately :(

**Comment by luisacicolini** [comment]
I tried to clean up the proofs and make them prettier/shorter and could not find any more simplifications beyond the current state - suggestions to make them better are very welcome :)

**Pull request merged**